docs/syntax-key-destructuring.md

```diff
@@ -0,0 +1,182 @@
+# Key destructuring
+
+## Summary
+
+Introduce a new syntax for unpacking key values into their own variables, such that:
+
+```lua
+local { .a, .b } = t
+-- a == t.a
+-- b == t.b
+```
+
+## Motivation
+
+Simple indexes on tables are very common both in and outside of Luau. A common use case is large libraries. It is common in the web world to see something like:
+
+```js
+const { useState, useEffect } = require("react");
+```
+
+...which allows you to quickly use `useState` and `useEffect` without fully qualifying it in the form of `React.useState` and `React.useEffect`. In Luau, if you do not want to fully qualify common React functions, the top of your file will often look like:
+
+```lua
+local useEffect = React.useEffect
+local useMemo = React.useMemo
+local useState = React.useState
+-- etc
+```
+
+...which creates a lot of redundant cruft.
+
+It is also common to want to have short identifiers to React properties, which basically always map onto a variable of the same name. As an anecdote, a regex search of `^\s+local (\w+) = \w+\.\1$` comes up 103 times in the My Movie codebase, many in the form of indexing React properties:
+
+```lua
+local position = props.position
+local style = props.style
+-- etc...
+```
+
+...whereas in JavaScript this would look like:
+```js
+const { position, style } = props
+
+// Or even...
+function MyComponent({
+	position,
+	style,
+})
+```
+
+React properties are themselves an example of a common idiom of passing around large tables as function arguments, such as with HTTP requests:
+
+```js
+// JavaScript
+get("/users", ({
+	users,
+	nextPageCursor,
+}) => { /* code */ })
+```
+
+## Design
+
+This RFC proposes expanding the grammar of `binding`:
+```patch
+binding = NAME [':' Type]
++         | `{` keydestructor [`,` keydestructor] [`,`] `}` [':' Type]
```

Comment:
This is supposed to be zero or more `, keydestructor` I just am blanking on how to write that at the moment

Reply:
Reminder that braces exists. (`{some grammar}` means parse that in a loop)

Reply:
I guess it would realistically get parsed like this but more statefully?

```
| '{' [keydestructors1] '}' [':' Type]

keydestructors1 = keydestructor [keydestructors2]
keydestructors2 = ',' [keydestructors1]
```

It's more human-readable like this though.

```
| '{' {keydestructor ','} [keydestructor] '}' [':' Type]
```

Reply:
I thought `{}` was one-or-more, sounds good.